tests/test_matrix_auth_discovery.py

```
@@ -0,0 +1,158 @@
+    client.device_id = None
+    client.restore_login = MagicMock()
+    client.whoami = AsyncMock()
+    return client
+
+
+@pytest.mark.asyncio
+async def test_whoami_discovers_missing_device_id(mock_client):
+    """Test that whoami is called to discover missing device_id and credentials are updated."""
+    auth_info = MatrixAuthInfo(
+        homeserver="https://matrix.org",
+        access_token="test_token",
```

Ruff (0.14.14)

[error] Possible hardcoded password assigned to argument: "access_token"

(S106)